### PR TITLE
ENH: stateless supervisor and worker

### LINKED ISSUE #316: BUG: concurrent operations cause error
### Describe the bug
When `xinference list` and `xinference launch` are executed at the same time, an error occured.

### To Reproduce
To help us to reproduce this bug, please provide information below:

1. Your Python version.
2. The version of xinference you use.
3. Versions of crucial packages.
4. Full stack of the error.
```python
ERROR:xinference.core.supervisor:Worker timeout. address: 0.0.0.0:29566, influenced models: []
llama_new_context_with_model: kv self size  =  640.00 MB
ERROR:xinference.core.restful_api:<asyncio.locks.Lock object at 0x7f1f2f794fa0 [locked, waiters:1]> is bound to a different event loop
Traceback (most recent call last):
  File "/mnt/data/data_wangzhong/repos/inference/xinference/core/restful_api.py", line 330, in launch_model
    await self._supervisor_ref.launch_builtin_model(
  File "xoscar/core.pyx", line 270, in __pyx_actor_method_wrapper
  File "xoscar/core.pyx", line 404, in _handle_actor_result
  File "xoscar/core.pyx", line 447, in _run_actor_async_generator
  File "/home/wangzhong/data/miniconda3/lib/python3.10/asyncio/locks.py", line 14, in __aenter__
    await self.acquire()
  File "/home/wangzhong/data/miniconda3/lib/python3.10/asyncio/locks.py", line 106, in acquire
    fut = self._get_loop().create_future()
  File "/home/wangzhong/data/miniconda3/lib/python3.10/asyncio/mixins.py", line 30, in _get_loop
    raise RuntimeError(f'{self!r} is bound to a different event loop')
RuntimeError: <asyncio.locks.Lock object at 0x7f1f2f794fa0 [locked, waiters:1]> is bound to a different event loop
ERROR:xinference.core.restful_api:dictionary changed size during iteration
Traceback (most recent call last):
  File "/mnt/data/data_wangzhong/repos/inference/xinference/core/restful_api.py", line 286, in list_models
    return await self._supervisor_ref.list_models()
  File "xoscar/core.pyx", line 266, in __pyx_actor_method_wrapper
  File "xoscar/core.pyx", line 269, in xoscar.core.__pyx_actor_method_wrapper
  File "/mnt/data/data_wangzhong/repos/inference/xinference/core/utils.py", line 25, in wrapped
    ret = await func(*args, **kwargs)
  File "/mnt/data/data_wangzhong/repos/inference/xinference/core/supervisor.py", line 164, in list_models
    for worker in self._worker_address_to_worker.values():
RuntimeError: dictionary changed size during iteration
```
5. Minimized code to reproduce the error.

### PR DESCRIPTION
**Pull request by UranusSeven**
This update addresses the following issues:

- Requests being blocked while another operation is in progress.
- Chat interface not responding in a streaming manner.
- Web UI redirects to the chat interface page even when the launch fails.

Resolves #316 #502 #516 #541 #540

### CODE REVIEW

File: xinference/core/chat_interface.py [hunk 1]
```diff
@@ -61,61 +75,67 @@ def build(self) -> "gr.Blocks":
...
 
         return gr.ChatInterface(
             fn=generate_wrapper,
-            # TODO: Change min/max based on model context size
             additional_inputs=[
                 gr.Slider(
-                    minimum=1, maximum=2048, value=1024, step=1, label="Max Tokens"
+                    minimum=1,
+                    maximum=self.context_length,
+                    value=128,
```
Comment (aresnow1) [inline xinference/core/chat_interface.py:130]: 128 is too small for chat.

File: xinference/core/supervisor.py [hunk 2]
```diff
@@ -86,7 +90,10 @@ async def _choose_worker(self) -> xo.ActorRefType["WorkerActor"]:
         # TODO: better allocation strategy.
         min_running_model_count = None
         target_worker = None
-        for worker in self._worker_address_to_worker.values():
+        async with self._lock:
+            workers = list(self._worker_address_to_worker.values())
```
Comment (codingl2k1) [inline xinference/core/supervisor.py:94]: Here, the lock is not required. All operations on the dictionary are synchronous, and any operation on the dictionary will not be interrupted by async calls.

File: xinference/core/supervisor.py [hunk 3]
```diff
@@ -183,7 +197,10 @@ async def register_model(self, model_type: str, model: str, persist: bool):
             register_llm(llm_family, persist)
 
             if not self.is_local_deployment:
-                for worker in self._worker_address_to_worker.values():
+                async with self._lock:
+                    workers = list(self._worker_address_to_worker.values())
```
Comment (codingl2k1) [inline xinference/core/supervisor.py:201]: Ditto

File: xinference/core/supervisor.py [hunk 4]
```diff
@@ -196,7 +213,10 @@ async def unregister_model(self, model_type: str, model_name: str):
             unregister_llm(model_name)
 
             if not self.is_local_deployment:
-                for worker in self._worker_address_to_worker.values():
+                async with self._lock:
+                    workers = list(self._worker_address_to_worker.values())
```
Comment (codingl2k1) [inline xinference/core/supervisor.py:217]: Ditto

File: xinference/core/supervisor.py [hunk 5]
```diff
@@ -330,42 +350,44 @@ async def _launch_one_model(_replica_model_uid):
...
-                        address,
-                        dead_models,
-                    )
-                    dead_nodes.append(address)
-
-            for address in dead_nodes:
-                self._worker_status.pop(address)
-                self._worker_address_to_worker.pop(address)
+            async with self._lock:
+                for address, status in self._worker_status.items():
```
Comment (codingl2k1) [inline xinference/core/supervisor.py:354]: Ditto. But be aware that we can not await inside the for loop.

File: xinference/core/supervisor.py [hunk 6]
```diff
@@ -376,42 +398,45 @@ async def _terminate_one_model(_replica_model_uid):
...
         info = await worker_ref.describe_model(model_uid=replica_model_uid)
         info["replica"] = replica_info.replica
         return info
 
     @log_async(logger=logger)
     async def list_models(self) -> Dict[str, Dict[str, Any]]:
         ret = {}
-        for worker in self._worker_address_to_worker.values():
+        async with self._lock:
+            workers = list(self._worker_address_to_worker.values())
```
Comment (codingl2k1) [inline xinference/core/supervisor.py:437]: Ditto

File: xinference/core/worker.py [hunk 7]
```diff
@@ -74,17 +76,29 @@ async def __pre_destroy__(self):
...
+    async def allocate_devices(self, model_uid: str, n_gpu: int) -> List[int]:
         if n_gpu > len(self._total_cuda_devices) - len(self._gpu_to_model_uid):
             raise RuntimeError("No available slot found for the model")
+
         devices: List[int] = [
             dev for dev in self._total_cuda_devices if dev not in self._gpu_to_model_uid
         ][:n_gpu]
+        async with self._lock:
+            for dev in devices:
+                self._gpu_to_model_uid[int(dev)] = model_uid
```
Comment (codingl2k1) [inline xinference/core/worker.py:88]: Ditto. But be aware that we can not await inside the for loop.

File: xinference/core/worker.py [hunk 8]
```diff
@@ -74,17 +76,29 @@ async def __pre_destroy__(self):
...
+
         return sorted(devices)
 
+    async def release_devices(self, model_uid: str):
+        async with self._lock:
+            devices = [
```
Comment (codingl2k1) [inline xinference/core/worker.py:94]: Ditto. But be aware that we can not await inside the for loop.

File: xinference/core/worker.py [hunk 9]
```diff
@@ -267,26 +281,29 @@ async def terminate_model(self, model_uid: str):
...
         del self._model_uid_to_addr[model_uid]
 
-    @log_sync(logger=logger)
-    def list_models(self) -> Dict[str, Dict[str, Any]]:
+    @log_async(logger=logger)
+    async def list_models(self) -> Dict[str, Dict[str, Any]]:
         ret = {}
-        for k, v in self._model_uid_to_model_spec.items():
+        async with self._lock:
+            items = list(self._model_uid_to_model_spec.items())
```
Comment (codingl2k1) [inline xinference/core/worker.py:288]: Ditto

File: xinference/core/chat_interface.py [hunk 10]
```diff
@@ -12,42 +12,56 @@
...
+            interface = self.build_chat_interface()
         else:
-            interface = self.build_generate_interface(
-                model,
-                model_info,
-            )
+            interface = self.build_generate_interface()
 
+        interface.queue()
+        interface.startup_events()
```
Comment (ChengjieLi28) [inline xinference/core/chat_interface.py:64]: Plz add a comment here to indicate why should use `startup_events` method for better reading. Link to that issue.

File: xinference/core/chat_interface.py (hunk 1 above)
Comment (UranusSeven) [reply xinference/core/chat_interface.py:133]: This is applicable to the 'generate' function, not the 'chat' function. In this scenario, generating a large number of tokens at once may not be advisable as the generated content needs frequent updating.

**Review by aresnow1 — APPROVED** [review]
LGTM

**Pull request merged**